File: stack/max_valid_parentheses.py

def count_longest_valid_length(string: str) -> int:  # LeetCode Q.32.
    isolate_indices: list[int] = []  # Stack for indices of unpaired left parentheses.
    completions: list[list[int]] = []  # Format: [start idx, end idx].

    max_len = 0
    for idx, parenthesis in enumerate(string):
        if parenthesis == ")":
            if not isolate_indices:  # Completions are stopped by an unpaired right parenthesis.
                completions.clear()
                continue

            pair_idx = isolate_indices.pop(-1)
            while completions and pair_idx < completions[-1][0] < completions[-1][1] < idx:
                completions.pop(-1)  # Merge current pair with past completions.

            if completions and completions[-1][1] + 1 == pair_idx:  # Extend to last completion.
                completions[-1][1] = idx

            else:  # Current pair becomes last completion.
                completions.append([pair_idx, idx])

            current_len = completions[-1][1] + 1 - completions[-1][0]
            if current_len > max_len:
                max_len = current_len
            continue

        isolate_indices.append(idx)

    return max_len

File: stack/0032_hard/max_valid_parentheses.cpp
#include <string>
#include <stack>
using namespace std;

int count_longest_valid_length(string s) // LeetCode Q.32.
{
    stack<int> open_parentheses; // Indices of open parentheses.
    // Format: {open idx, close idx, parentheses len}.
    stack<tuple<int, int, int>> complete_parentheses;

    int longest_valid_len = 0;
    for (int idx = 0; idx < s.length(); idx++)
    {
        if (s[idx] == '(')
        {
            open_parentheses.push(idx);
            continue;
        }

        if (!open_parentheses.empty())
        {
            int open_idx = open_parentheses.top(), length = 2;
            open_parentheses.pop();

            if (!complete_parentheses.empty())
            {
                auto [prev_open_idx, prev_close_idx, prev_len] = complete_parentheses.top();
                if (open_idx < prev_open_idx && prev_close_idx < idx)
                {
                    length += prev_len;
                    complete_parentheses.pop();
                }
            }

            if (!complete_parentheses.empty())
            {
                auto [prev_open_idx, prev_close_idx, prev_len] = complete_parentheses.top();
                if (prev_close_idx + 1 == open_idx)
                {
                    open_idx = prev_open_idx;
                    length += prev_len;
                    complete_parentheses.pop();
                }
            }

            if (length > longest_valid_len)
            {
                longest_valid_len = length;
            }
            complete_parentheses.push({open_idx, idx, length});
        }
    }

    return longest_valid_len;
}
